src/core/dappletContext.ts

```diff
@@ -20,20 +21,70 @@ export class DappletContext {
 
     constructor(config: ContextConfig = DEFAULT_CONFIG) {
         config = { ...DEFAULT_CONFIG, ...config };
-        this.dappletProviders = config.providers!;
+        this.dappletProviders = config.providers! || [];
         this.featureRegistry = new FeatureRegistry(config.features || []);
     }
 
     async processRequest(request: DappletRequest): Promise<DappletTxResult> {
+        // dapplet loading and validation
+        for (let i = 0; i < request.frames.length; i++) {
+            const dapplet = await this._loadDapplet(request.frames[i].dappletId);
+            this._validateDapplet(dapplet);
+            request.frames[i].dapplet = dapplet;
+        }
+
         const engine = new DappletEngine(request, this);
-        engine.onStatusChanged((statuses) => console.log(statuses.map(s => FrameStatus[s]).join(", ")));
-        await engine.load();
-        engine.validate();
-        
+        engine.run();
+
         //const activity = new DappletActivity(request, this);
         return {};
     }
 
+    private async _loadDapplet(dappletId: string): Promise<DappletTemplate> {
+        for (const provider of this.dappletProviders) {
+            try {
+                const dapplet = await provider.loadDapplet(dappletId);
```

Comment:
we need probably to specify either the dappletId or provide a dappletIdType.
consider dappletId is a swarm-hash. it can be natively used to retrieve an object from swarm provider, but HttpProvider may need to know that it is a swarm-hash in order to access the swarm-fallback-server.

Reply:
As a separated class or just simple type?